activerecord/test/cases/adapters/mysql2/virtual_column_test.rb

```diff
@@ -0,0 +1,59 @@
+require "cases/helper"
+require "support/schema_dumping_helper"
+
+if ActiveRecord::Base.connection.supports_virtual_columns?
+  class Mysql2VirtualColumnTest < ActiveRecord::Mysql2TestCase
+    include SchemaDumpingHelper
+
+    self.use_transactional_tests = false
+
+    class VirtualColumn < ActiveRecord::Base
+    end
+
+    def setup
+      @connection = ActiveRecord::Base.connection
+      @connection.create_table :virtual_columns, force: true do |t|
+        t.string  :name
+        t.virtual :upper_name,  type: :string,  as: "UPPER(name)"
+        t.virtual :name_length, type: :integer, as: "LENGTH(name)", stored: true
+      end
+      VirtualColumn.create(name: "Rails")
+    end
+
+    def teardown
+      @connection.drop_table :virtual_columns, if_exists: true
+      VirtualColumn.reset_column_information
+    end
+
+    def test_virtual_column
+      column = VirtualColumn.columns_hash["upper_name"]
+      assert column.virtual?
+      assert_equal "VIRTUAL GENERATED", column.extra
+      assert_equal "RAILS", VirtualColumn.take.upper_name
+    end
+
+    def test_stored_column
+      column = VirtualColumn.columns_hash["name_length"]
+      assert column.virtual?
+      assert_equal "STORED GENERATED", column.extra
+      assert_equal 5, VirtualColumn.take.name_length
+    end
+
+    def test_change_table
+      @connection.change_table :virtual_columns do |t|
+        t.virtual :lower_name, type: :string, as: "LOWER(name)"
+      end
+      VirtualColumn.reset_column_information
+      column = VirtualColumn.columns_hash["lower_name"]
+      assert column.virtual?
+      assert_equal "VIRTUAL GENERATED", column.extra
+      assert_equal "rails", VirtualColumn.take.lower_name
+    end
+
+    def test_schema_dumping
```

Comment:
This test fails with these two MySQL servers.

- Server version: 8.0.0-dmr MySQL Community Server (GPL)
- Server version: 5.7.17 MySQL Community Server (GPL)

```ruby
$ ARCONN=mysql2 bundle exec ruby -Itest test/cases/adap                                        ters/mysql2/virtual_column_test.rb -n test_schema_dumping

Using mysql2
Run options: -n test_schema_dumping --seed 61883

# Running:

F

Finished in 0.835548s, 1.1968 runs/s, 2.3936 assertions/s.

  1) Failure:
Mysql2VirtualColumnTest#test_schema_dumping [test/cases/adapters/mysql2/virtual_column_test.rb:55]:
Expected /t\.virtual\s+"upper_name",\s+type: :string,\s+as: "UPPER\(name\)"$/ to match "# This file is auto-generated from the current state of the database. Instead\n# of editing this file, please use the migrations feature of Active Record to\n# incrementally modify your database, and then regenerate this schema definition.\n#\n# Note that this schema.rb definition is the authoritative source for your\n# database schema. If you need to create the application database on another\n# system, you should be using db:schema:load, not running all the migrations\n# from scratch. The latter is a flawed and unsustainable approach (the more migrations\n# you'll amass, the slower it'll run and the greater likelihood for issues).\n#\n# It's strongly recommended that you check this file into your version control system.\n\nActiveRecord::Schema.define(version: 0) do\n\n  create_table \"virtual_columns\", force: :cascade, options: \"ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci\" do |t|\n    t.string \"name\"\n    t.virtual \"upper_name\", type: :string, as: \"upper(`name`)\"\n    t.virtual \"name_length\", type: :integer, as: \"length(`name`)\", stored: true\n  end\n\nend\n".

1 runs, 2 assertions, 1 failures, 0 errors, 0 skips
$
```

It is likely due to UPPER and LENGTH function are converted into lowercase upper and length at MySQL server side.

```sql
mysql> CREATE TABLE `virtual_columns` (`id` bigint AUTO_INCREMENT PRIMARY KEY, `name` varchar(255), `upper_name` varchar(255) AS (UPPER(name)), `name_length` int AS (LENGTH(name)) STORED) ENGINE=InnoDB;
Query OK, 0 rows affected (0.00 sec)

mysql> show create table `virtual_columns`\G
*************************** 1. row ***************************
       Table: virtual_columns
Create Table: CREATE TABLE `virtual_columns` (
  `id` bigint(20) NOT NULL AUTO_INCREMENT,
  `name` varchar(255) DEFAULT NULL,
  `upper_name` varchar(255) GENERATED ALWAYS AS (upper(`name`)) VIRTUAL,
  `name_length` int(11) GENERATED ALWAYS AS (length(`name`)) STORED,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=latin1
1 row in set (0.00 sec)

mysql>
```